x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/TransportLoadTrainedModelPackage.java

```diff
@@ -0,0 +1,143 @@
+
+public class TransportLoadTrainedModelPackage extends TransportMasterNodeAction<Request, AcknowledgedResponse> {
+
+    private static final int DEFAULT_CHUNK_SIZE = 4 * 1024 * 1024; // 4MB
+
+    private static final Logger logger = LogManager.getLogger(TransportLoadTrainedModelPackage.class);
+
+    private final Client client;
+
+    @Inject
+    public TransportLoadTrainedModelPackage(
+        TransportService transportService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        ActionFilters actionFilters,
+        IndexNameExpressionResolver indexNameExpressionResolver,
+        Client client
+    ) {
+        super(
+            LoadTrainedModelPackageAction.NAME,
+            transportService,
+            clusterService,
+            threadPool,
+            actionFilters,
+            LoadTrainedModelPackageAction.Request::new,
+            indexNameExpressionResolver,
+            NodeAcknowledgedResponse::new,
+            ThreadPool.Names.SAME
+        );
+        this.client = new OriginSettingClient(client, ML_ORIGIN);
+    }
+
+    @Override
+    protected void masterOperation(Task task, Request request, ClusterState state, ActionListener<AcknowledgedResponse> listener)
+        throws Exception {
+        ModelPackageConfig modelPackageConfig = request.getModelPackageConfig();
+        String repository = modelPackageConfig.getModelRepository();
+        String modelId = request.getModelId();
+        long size = modelPackageConfig.getSize();
+        String packagedModelId = modelPackageConfig.getPackagedModelId();
+
+        threadPool.executor(MachineLearningPackageLoader.UTILITY_THREAD_POOL_NAME).execute(() -> {
+            try {
+                URI uri = new URI(repository).resolve(packagedModelId + ModelLoaderUtils.MODEL_FILE_EXTENSION);
+                InputStream modelInputStream = ModelLoaderUtils.getInputStreamFromModelRepository(uri);
+
+                ModelLoaderUtils.InputStreamChunker chunkIterator = new ModelLoaderUtils.InputStreamChunker(
+                    modelInputStream,
+                    DEFAULT_CHUNK_SIZE
+                );
+
+                // simple round up
+                int totalParts = (int) ((size + DEFAULT_CHUNK_SIZE - 1) / DEFAULT_CHUNK_SIZE);
+
+                for (int part = 0; part < totalParts; ++part) {
+                    BytesArray definition = chunkIterator.next();
+                    PutTrainedModelDefinitionPartAction.Request r = new PutTrainedModelDefinitionPartAction.Request(
+                        modelId,
+                        definition,
+                        part,
+                        size,
+                        totalParts
+                    );
+
+                    client.execute(PutTrainedModelDefinitionPartAction.INSTANCE, r).actionGet();
+                }
+                logger.trace(() -> format("finished uploading model using [%d] parts", totalParts));
```

Comment:
This and the other trace level logs could be debug or info. The download only happens once and is significant it warrants logging at a higher level